cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -0,0 +1,239 @@
+  ASSERT_OK_AND_ASSIGN(Datum actual_microsecond, Microsecond(time_points));
+  ASSERT_OK_AND_ASSIGN(Datum actual_nanosecond, Nanosecond(time_points));
+
+  ASSERT_EQ(actual_year, year);
+  ASSERT_EQ(actual_month, month);
+  ASSERT_EQ(actual_day, day);
+  ASSERT_EQ(actual_day_of_year, day_of_year);
+  ASSERT_EQ(actual_week, week);
+  ASSERT_EQ(actual_quarter, quarter);
+  ASSERT_EQ(actual_day_of_week, day_of_week);
+  ASSERT_EQ(actual_hour, hour);
+  ASSERT_EQ(actual_minute, minute);
+  ASSERT_EQ(actual_second, second);
+  ASSERT_EQ(actual_millisecond, millisecond);
+  ASSERT_EQ(actual_microsecond, microsecond);
+  ASSERT_EQ(actual_nanosecond, nanosecond);
+
+  CheckScalarUnary("year", time_points, year);
+  CheckScalarUnary("month", time_points, month);
+  CheckScalarUnary("day", time_points, day);
+  CheckScalarUnary("day_of_year", time_points, day_of_year);
+  CheckScalarUnary("week", time_points, week);
+  CheckScalarUnary("quarter", time_points, quarter);
+  CheckScalarUnary("day_of_week", time_points, day_of_week);
+  CheckScalarUnary("hour", time_points, hour);
+  CheckScalarUnary("minute", time_points, minute);
+  CheckScalarUnary("second", time_points, second);
+  CheckScalarUnary("millisecond", time_points, millisecond);
+  CheckScalarUnary("microsecond", time_points, microsecond);
+  CheckScalarUnary("nanosecond", time_points, nanosecond);
+}
+
+TEST(ScalarTemporalTest, TestTemporalComponentExtraction) {
+  const char* json_second = "[59, 951866603, -2208981640, 2000000000]";
+  const char* json_milli = "[59000, 951866603000, -2208981640000, 2000000000000]";
+  const char* json_micro =
+      "[59000000, 951866603000000, -2208981640000000, 2000000000000000]";
+  const char* json_nano =
+      "[59000000000, 951866603000000000, -2208981640000000000, 2000000000000000000]";
+
+  auto time_points_second = ArrayFromJSON(timestamp(TimeUnit::SECOND), json_second);
+  auto time_points_milli = ArrayFromJSON(timestamp(TimeUnit::MILLI), json_milli);
+  auto time_points_micro = ArrayFromJSON(timestamp(TimeUnit::MICRO), json_micro);
+  auto time_points_nano = ArrayFromJSON(timestamp(TimeUnit::NANO), json_nano);
+
+  auto year = ArrayFromJSON(int64(), "[1970, 2000, 1900, 2033]");
+  auto month = ArrayFromJSON(int64(), "[1, 2, 1, 5]");
+  auto day = ArrayFromJSON(int64(), "[1, 29, 1, 18]");
+  auto day_of_year = ArrayFromJSON(int64(), "[1, 60, 1, 138]");
+  auto week = ArrayFromJSON(int64(), "[1, 9, 1, 20]");
+  auto quarter = ArrayFromJSON(int64(), "[1, 1, 1, 2]");
+  auto day_of_week = ArrayFromJSON(int64(), "[4, 2, 1, 3]");
+  auto hour = ArrayFromJSON(int64(), "[0, 23, 1, 3]");
+  auto minute = ArrayFromJSON(int64(), "[0, 23, 59, 33]");
+  auto second = ArrayFromJSON(int64(), "[59, 23, 20, 20]");
+  auto millisecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto microsecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
+  auto nanosecond = ArrayFromJSON(int64(), "[0, 0, 0, 0]");
```

Comment:
Hmm... so you never test with a non-zero number of milli/micro/nanoseconds?
Also, how about testing with null values as well?